akka-cluster-sharding/src/main/scala/akka/cluster/sharding/ClusterSharding.scala

```diff
@@ -335,6 +334,154 @@ class ClusterSharding(system: ExtendedActorSystem) extends Extension {
+   *   entity from the incoming message, if the partial function does not match the message will
+   *   be `unhandled`, i.e. posted as `Unhandled` messages on the event stream
+   * @param extractShardId function to determine the shard id for an incoming message, only messages
+   *   that passed the `extractEntityId` will be used
+   * @param allocationStrategy possibility to use a custom shard allocation and
+   *   rebalancing logic
+   * @param handOffStopMessage the message that will be sent to entities when they are to be stopped
+   *   for a rebalance or graceful shutdown of a `ShardRegion`, e.g. `PoisonPill`.
+   * @return the actor ref of the [[ShardRegion]] that is to be responsible for the shard
+   */
+  def start(
+    typeName:           String,
+    propsGen:           ShardRegion.PropsGenerator,
+    settings:           ClusterShardingSettings,
+    extractEntityId:    ShardRegion.ExtractEntityId,
+    extractShardId:     ShardRegion.ExtractShardId,
+    allocationStrategy: ShardAllocationStrategy,
+    handOffStopMessage: Any): ActorRef = {
+
+    requireClusterRole(settings.role)
+    implicit val timeout = system.settings.CreationTimeout
+    val startMsg = Start(typeName, propsGen, settings,
+      extractEntityId, extractShardId, allocationStrategy, handOffStopMessage)
+    val Started(shardRegion) = Await.result(guardian ? startMsg, timeout.duration)
+    regions.put(typeName, shardRegion)
+    shardRegion
+  }
+
+  /**
+   * Scala API: Register a named entity type by defining a factory for the [[akka.actor.Props]] of
+   * the entity actor and functions to extract entity and shard identifier from messages. The
+   * [[ShardRegion]] actor for this type can later be retrieved with the [[#shardRegion]] method.
+   *
+   * The default shard allocation strategy [[ShardCoordinator.LeastShardAllocationStrategy]]
+   * is used. [[akka.actor.PoisonPill]] is used as `handOffStopMessage`.
+   *
+   * Some settings can be configured as described in the `akka.cluster.sharding` section
+   * of the `reference.conf`.
+   *
+   * @param typeName the name of the entity type
+   * @param propsGen function that, given an entity id, returns the `Props` of the entity actors
+   *  that will be created by the `ShardRegion`
+   * @param settings configuration settings, see [[ClusterShardingSettings]]
+   * @param extractEntityId partial function to extract the entity id and the message to send to the
+   *   entity from the incoming message, if the partial function does not match the message will
+   *   be `unhandled`, i.e. posted as `Unhandled` messages on the event stream
+   * @param extractShardId function to determine the shard id for an incoming message, only messages
+   *   that passed the `extractEntityId` will be used
+   * @return the actor ref of the [[ShardRegion]] that is to be responsible for the shard
+   */
+  def start(
+    typeName:        String,
+    propsGen:        ShardRegion.PropsGenerator,
```

Comment:
I don't like it that it is called Factory in Java and Generator in Scala, could we just skip the function type alias instead, and call the param propsFactory here as well?

Reply:
Sorry, I saw the differences between the MessageExtractor and the Scala function types and didn't realize that was because there's 2 functions in that interface. I'll change it.